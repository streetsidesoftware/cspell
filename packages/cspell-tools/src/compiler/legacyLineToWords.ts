import { opConcatMap, opFilter, opMap, pipe } from '@cspell/cspell-pipe/sync';

import { regExpSpaceOrDash, splitCamelCaseIfAllowed } from './splitCamelCaseIfAllowed.js';
import type { AllowedSplitWordsCollection } from './WordsCollection.js';

const regNonWord = /[^\p{L}\p{M}' _\d]+/giu;
const regExpRepeatChars = /(.)\1{5}/i;

export function legacyLineToWords(
    line: string,
    keepCase: boolean,
    allowedSplitWords: AllowedSplitWordsCollection
): Iterable<string> {
    // Remove punctuation and non-letters.
    const filteredLine = line.replace(regNonWord, '|');
    const wordGroups = filteredLine.split('|');

    const words = pipe(
        wordGroups,
        opConcatMap((a) => [...a.split(regExpSpaceOrDash)]),
        opConcatMap((a) => splitCamelCaseIfAllowed(a, allowedSplitWords, keepCase)),
        opMap((a) => a.trim()),
        opFilter((a) => !!a),
        opFilter((s) => !regExpRepeatChars.test(s))
    );

    return words;
}

export function* legacyLinesToWords(
    lines: Iterable<string>,
    keepCase: boolean,
    allowedSplitWords: AllowedSplitWordsCollection
): Iterable<string> {
    for (const line of lines) {
        yield* legacyLineToWords(line, keepCase, allowedSplitWords);
    }
}
