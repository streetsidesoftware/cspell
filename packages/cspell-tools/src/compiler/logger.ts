// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Logger = (message?: any, ...optionalParams: any[]) => void;

const defaultLogger = console.log;
let log: Logger = defaultLogger;

export function setLogger(logger?: Logger): void {
    log = logger ?? defaultLogger;
}

export function getLogger(): Logger {
    return log;
}
