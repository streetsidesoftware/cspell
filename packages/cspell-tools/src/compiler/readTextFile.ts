import * as fs from 'fs-extra';
import * as zlib from 'zlib';

export function readTextFile(filename: string): Promise<string> {
    const content = fs
        .readFile(filename)
        .then((buffer) => (/\.gz$/.test(filename) ? zlib.gunzipSync(buffer) : buffer))
        .then((buffer) => buffer.toString('utf8'));
    return content;
}

export async function readTextFileLines(filename: string): Promise<string[]> {
    const content = await readTextFile(filename);
    return content.split('\n');
}
