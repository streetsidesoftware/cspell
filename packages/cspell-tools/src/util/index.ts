export function isDefined<T>(a: T | undefined | null): a is T {
    return !!a || (a !== null && a !== undefined);
}
