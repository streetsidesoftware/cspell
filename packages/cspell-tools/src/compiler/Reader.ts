import { pipe } from '@cspell/cspell-pipe/sync';
import {
    CASE_INSENSITIVE_PREFIX,
    COMPOUND_FIX,
    createDictionaryLineParser,
    FORBID_PREFIX,
    importTrie,
    Trie,
} from 'cspell-trie-lib';
import * as HR from 'hunspell-reader';
import { AffWord } from 'hunspell-reader';
import { legacyLinesToWords } from './legacyLineToWords';
import { readTextFileLines } from './readTextFile';

const regHunspellFile = /\.(dic|aff)$/i;

export interface ReaderOptions {
    /**
     * Max Hunspell recursive depth.
     */
    maxDepth?: number;
    /**
     * split words if necessary.
     */
    splitWords: boolean;
    /**
     * Generate a case insensitive version.
     */
    generateNonStrictAlternatives: boolean;
    /**
     * Indicate that it is an unformatted file and needs to be cleaned
     * before processing. Applies only to text file sources.
     * @default false
     */
    legacy?: boolean;
}

type ReaderFn = (filename: string, options: ReaderOptions) => Promise<BaseReader>;

// cspell:word dedupe
const DEDUPE_SIZE = 1000;

interface ReaderSelector {
    test: RegExp;
    method: ReaderFn;
}

export type AnnotatedWord = string;

interface BaseReader {
    size: number;
    words: Iterable<AnnotatedWord>;
}

export interface Reader extends BaseReader, Iterable<string> {}

// Readers first match wins
const readers: ReaderSelector[] = [
    { test: /\.trie\b/, method: trieFileReader },
    { test: regHunspellFile, method: readHunspellFiles },
];

function findMatchingReader(filename: string, options: ReaderOptions): Promise<BaseReader> {
    for (const reader of readers) {
        if (reader.test.test(filename)) {
            return reader.method(filename, options);
        }
    }
    return textFileReader(filename, options);
}

export async function createReader(filename: string, options: ReaderOptions): Promise<Reader> {
    const baseReader = await findMatchingReader(filename, options);
    return Object.assign(baseReader, {
        [Symbol.iterator]: () => baseReader.words[Symbol.iterator](),
    });
}

export async function readHunspellFiles(filename: string, options: ReaderOptions): Promise<BaseReader> {
    const dicFile = filename.replace(regHunspellFile, '.dic');
    const affFile = filename.replace(regHunspellFile, '.aff');

    const reader = await HR.IterableHunspellReader.createFromFiles(affFile, dicFile);
    reader.maxDepth = options.maxDepth !== undefined ? options.maxDepth : reader.maxDepth;

    const normalizeAndDedupe = opCompose(_stripCaseAndAccents, dedupeAndSort);
    const words = pipe(reader.seqAffWords(), _mapAffWords, normalizeAndDedupe);

    return {
        size: reader.dic.length,
        words,
    };
}

async function trieFileReader(filename: string): Promise<BaseReader> {
    const trieRoot = importTrie(await readTextFileLines(filename));
    const trie = new Trie(trieRoot);
    const words = trie.words();
    return {
        get size() {
            return trie.size();
        },
        words,
    };
}

async function textFileReader(filename: string, options: ReaderOptions): Promise<BaseReader> {
    const lines = await readTextFileLines(filename);
    const { splitWords: split, generateNonStrictAlternatives: stripCaseAndAccents } = options;

    const normalizeLines = createDictionaryLineParser({ stripCaseAndAccents, split });
    const parseLines = options.legacy ? opCompose(legacyLinesToWords, normalizeLines) : normalizeLines;
    const words = pipe(lines, parseLines, dedupeAndSort);

    return {
        size: lines.length,
        words,
    };
}

function* _stripCaseAndAccents(words: Iterable<AnnotatedWord>): Iterable<AnnotatedWord> {
    for (const word of words) {
        // Words are normalized to the compact format: e + ` => Ã¨
        yield word.normalize();
        // covert to lower case and strip accents.
        const n = word.toLowerCase().normalize('NFD').replace(/\p{M}/gu, '');
        // All words are added for case-insensitive searches.
        // It is a space / speed trade-off. In this case, speed is more important.
        yield CASE_INSENSITIVE_PREFIX + n;
    }
}

function* dedupeAndSort(words: Iterable<AnnotatedWord>): Iterable<AnnotatedWord> {
    const buffer: AnnotatedWord[] = [];

    function sortDedupeClear() {
        const s = new Set(buffer.sort());
        buffer.length = 0;
        return s;
    }

    for (const word of words) {
        buffer.push(word);
        if (buffer.length >= DEDUPE_SIZE) {
            yield* sortDedupeClear();
        }
    }
    yield* sortDedupeClear();
}

function* _mapAffWords(affWords: Iterable<AffWord>): Iterable<AnnotatedWord> {
    const hasSpecial = /[~+!]/;
    for (const affWord of affWords) {
        const { word, flags } = affWord;
        // For now do not include words with special characters.
        if (hasSpecial.test(word)) continue;
        const compound = flags.isCompoundForbidden ? '' : COMPOUND_FIX;
        const forbid = flags.isForbiddenWord ? FORBID_PREFIX : '';
        if (!forbid) {
            if (flags.canBeCompoundBegin || flags.isCompoundPermitted) yield word + compound;
            if (flags.canBeCompoundEnd || flags.isCompoundPermitted) yield compound + word;
            if (flags.canBeCompoundMiddle || flags.isCompoundPermitted) yield compound + word + compound;
            if (!flags.isOnlyAllowedInCompound) yield word;
        } else {
            yield forbid + word;
        }
    }
}

function opCompose<T>(...ops: ((i: Iterable<T>) => Iterable<T>)[]): (i: Iterable<T>) => Iterable<T> {
    return (i: Iterable<T>) => {
        for (const op of ops) {
            i = op(i);
        }
        return i;
    };
}

export const __testing__ = {
    _stripCaseAndAccents,
};
