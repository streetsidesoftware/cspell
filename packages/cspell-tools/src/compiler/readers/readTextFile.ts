import { promises as fs } from 'fs';
import { promisify } from 'util';
import * as zlib from 'zlib';

const gunzip = promisify(zlib.gunzip);

const isGzFile = /\.gz$/;

export function readTextFile(filename: string): Promise<string> {
    const content = fs
        .readFile(filename)
        .then((buffer) => (isGzFile.test(filename) ? gunzip(buffer) : buffer))
        .then((buffer) => buffer.toString('utf8'));
    return content;
}

export async function readTextFileLines(filename: string): Promise<string[]> {
    const content = await readTextFile(filename);
    return content.split('\n');
}
