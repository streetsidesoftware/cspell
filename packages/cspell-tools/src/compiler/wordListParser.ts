import { opCombine, opFilter, Operator } from '@cspell/cspell-pipe/sync';
import { uniqueFilter } from 'hunspell-reader/dist/util';
import { CompileOptions } from './CompileOptions';

export function createSortAndFilterOperation(options: CompileOptions): Operator<string> {
    const operations: Operator<string>[] = [
        options.stripNonStrictPrefix ? opFilter<string>((w) => !w.startsWith('~')) : undefined,
        opFilter<string>((a) => !!a),
        options.sort ? createInlineBufferedSort() : undefined,
        opFilter<string>(uniqueFilter(10000)),
    ].filter(isDefined);
    return opCombine(...operations);
}

function isDefined<T>(v: T | undefined): v is T {
    return v !== undefined;
}

function createInlineBufferedSort(bufferSize = 1000): (lines: Iterable<string>) => Iterable<string> {
    function* inlineBufferedSort(lines: Iterable<string>): Iterable<string> {
        const buffer: string[] = [];

        for (const line of lines) {
            buffer.push(line);
            if (buffer.length >= bufferSize) {
                buffer.sort();
                yield* buffer;
                buffer.length = 0;
            }
        }

        buffer.sort();
        yield* buffer;
    }

    return inlineBufferedSort;
}
