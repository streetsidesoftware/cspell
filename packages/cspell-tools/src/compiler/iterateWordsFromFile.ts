import { createReader, ReaderOptions } from './Reader';

export async function streamWordsFromFile(filename: string, options: ReaderOptions): Promise<Iterable<string>> {
    const reader = await createReader(filename, options);
    return reader;
}
