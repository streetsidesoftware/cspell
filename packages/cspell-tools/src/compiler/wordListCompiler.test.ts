// cSpell:ignore jpegs outing dirs lcode outring outrings

import { opConcatMap, opTake, pipe, toArray } from '@cspell/cspell-pipe/sync';
import * as Trie from 'cspell-trie-lib';
import { importTrie, isCircular, iteratorTrieWords, serializeTrie } from 'cspell-trie-lib';
import * as fsp from 'fs-extra';
import { uniqueFilter } from 'hunspell-reader/dist/util';
import * as path from 'path';
import { spyOnConsole } from '../test/console';
import { createTestHelper } from '../test/TestHelper';
import { streamWordsFromFile } from './iterateWordsFromFile';
import { legacyLineToWords } from './legacyLineToWords';
import { setLogger } from './logger';
import { ReaderOptions } from './Reader';
import { readTextFile } from './readTextFile';
import {
    compileTrie as _compileTrie,
    CompileTrieOptions,
    compileWordList as _compileWordList,
    __testing__,
} from './wordListCompiler';
import { createSortAndFilterOperation } from './wordListParser';

const testHelper = createTestHelper(__filename);

const samples = path.join(testHelper.packageRoot, '../Samples/dicts');
const sampleDictEnUS = path.join(samples, 'hunspell', 'en_US.dic');
const sampleDictEn = path.join(samples, 'en_US.txt');

const wordListHeader = __testing__.wordListHeader;

const { consoleOutput } = spyOnConsole();
setLogger(console.log);

const readOptionsAdjustCase: ReaderOptions = {
    splitWords: false,
    generateNonStrictAlternatives: true,
};

const readOptionsKeepCase: ReaderOptions = {
    splitWords: false,
    generateNonStrictAlternatives: false,
};

describe('Validate the wordListCompiler', () => {
    let temp = '.';
    beforeEach(() => {
        testHelper.cdToTempDir();
        temp = testHelper.resolveTemp();
        jest.resetAllMocks();
    });

    test.each`
        destFile
        ${'cities.txt'}
        ${'cities.txt.gz'}
    `('reading and normalizing to text file: $destFile', async ({ destFile }) => {
        const source = [...(await streamWordsFromFile(path.join(samples, 'cities.txt'), readOptionsKeepCase))];
        const destName = path.join(temp, destFile);
        await compileWordList(source, destName, { sort: false, stripNonStrictPrefix: true });
        const result = await readTextFile(destName);
        const expected = '\n# cspell-tools: keep-case no-split\n\n' + source.join('\n') + '\n';
        expect(result).toEqual(expected);
        expect(consoleOutput()).toMatchSnapshot();
    });

    test('compiling to a file without split', async () => {
        const source = await streamWordsFromFile(path.join(samples, 'cities.txt'), readOptionsAdjustCase);
        const destName = path.join(temp, 'cities2.txt');
        await compileWordList(source, destName, { sort: false, stripNonStrictPrefix: false });
        const output = await fsp.readFile(destName, 'utf8');
        expect(output).toBe(
            wordListHeader +
                '\n' +
                citiesSorted +
                citiesSorted
                    .toLowerCase()
                    .split('\n')
                    .filter((a) => !!a)
                    .map((a) => '~' + a)
                    .join('\n') +
                '\n'
        );
        expect(consoleOutput()).toMatchSnapshot();
    });

    test('tests normalized to a trie', () => {
        const words = citiesLegacyResult.split('\n');
        const nWords = toArray(legacyNormalizeWords(words));
        const tWords = [...Trie.iteratorTrieWords(normalizeWordsToTrie(words))];
        expect(tWords.sort()).toEqual([...new Set(nWords.sort())]);
        expect(consoleOutput()).toMatchSnapshot();
    });

    test.each`
        destFile
        ${'cities.trie'}
        ${'cities.trie.gz'}
    `('reading and normalizing to $destFile', async ({ destFile }) => {
        const source = [...(await streamWordsFromFile(path.join(samples, 'cities.txt'), readOptionsKeepCase))];
        const destName = path.join(temp, destFile);
        await compileTrie(source, destName, { sort: true, stripNonStrictPrefix: false });
        const resultFile = await readTextFile(destName);
        const resultLines = resultFile.split('\n');
        const node = Trie.importTrie(resultLines);
        const words = [...Trie.iteratorTrieWords(node)].sort();
        expect(words).toEqual(source.concat().sort());
        expect(consoleOutput()).toMatchSnapshot();
    });

    test('a simple hunspell dictionary depth 0', async () => {
        const source = await streamWordsFromFile(path.join(samples, 'hunspell', 'example.dic'), {
            ...readOptionsAdjustCase,
            maxDepth: 0,
        });
        const destName = path.join(temp, 'example0.txt');
        await compileWordList(source, destName, { sort: false, stripNonStrictPrefix: true });
        const output = await fsp.readFile(destName, 'utf8');
        expect(output).toBe(__testing__.wordListHeader + '\n' + 'hello\ntry\nwork\n');
        expect(consoleOutput()).toMatchSnapshot();
    });

    test('a simple hunspell dictionary depth 1', async () => {
        const source = await streamWordsFromFile(path.join(samples, 'hunspell/example.dic'), {
            ...readOptionsKeepCase,
            maxDepth: 1,
        });
        const destName = path.join(temp, 'example1.txt');
        await compileWordList(source, destName, { sort: false, stripNonStrictPrefix: true });
        const output = await fsp.readFile(destName, 'utf8');
        expect(output.split('\n')).toEqual(
            `\

            # cspell-tools: keep-case no-split

            hello
            rework
            tried
            try
            work
            worked
        `
                .split('\n')
                .map((line) => line.trim())
        );
        expect(consoleOutput()).toMatchSnapshot();
    });
});

describe('Validate Larger Dictionary', () => {
    test('en_US hunspell', async () => {
        const source = await streamWordsFromFile(sampleDictEnUS, readOptionsAdjustCase);
        const words = [...pipe(source, opTake(5000))];
        const trie = normalizeWordsToTrie(words);
        expect(isCircular(trie)).toBe(false);
        const nWords = toArray(legacyNormalizeWords(words)).sort().filter(uniqueFilter(1000));
        const results = iteratorTrieWords(trie).toArray().sort().filter(uniqueFilter(1000));
        expect(results).toEqual(nWords);
    }, 60000);

    test('en_US word list', async () => {
        const source = await streamWordsFromFile(sampleDictEn, readOptionsAdjustCase);
        const words = [...source];
        const trie = Trie.consolidate(normalizeWordsToTrie(words));
        expect(isCircular(trie)).toBe(false);
        const nWords = toArray(legacyNormalizeWords(words)).sort().filter(uniqueFilter(1000));
        const results = iteratorTrieWords(trie).toArray().sort();
        expect(results).toEqual(nWords);
        const data = serializeTrie(trie, { base: 40 });
        const trie2 = importTrie(data);
        const results2 = iteratorTrieWords(trie2).toArray();
        expect(results2).toEqual(results);
    }, 60000);
});

async function compileTrie(words: Iterable<string>, destFilename: string, options: CompileTrieOptions): Promise<void> {
    const normalizer = createSortAndFilterOperation(options);
    return _compileTrie(normalizer(words), destFilename, options);
}

async function compileWordList(
    lines: Iterable<string>,
    destFilename: string,
    options: CompileTrieOptions
): Promise<void> {
    const normalizer = createSortAndFilterOperation(options);
    return _compileWordList(normalizer(lines), destFilename, options);
}

function normalizeWordsToTrie(words: Iterable<string>): Trie.TrieRoot {
    return Trie.buildTrie(legacyNormalizeWords(words)).root;
}

function legacyNormalizeWords(lines: Iterable<string>): Iterable<string> {
    return pipe(
        lines,
        opConcatMap((line) => legacyLineToWords(line))
    );
}

// const cities = `\
// New York
// New Amsterdam
// Los Angeles
// San Francisco
// New Delhi
// Mexico City
// London
// Paris
// `;

const citiesSorted = `\
London
Los Angeles
Mexico City
New Amsterdam
New Delhi
New York
Paris
San Francisco
`;

const citiesLegacyResult = `\
new york
new
york
new amsterdam
amsterdam
los angeles
los
angeles
san francisco
san
francisco
new delhi
delhi
mexico city
mexico
city
london
paris
`;
