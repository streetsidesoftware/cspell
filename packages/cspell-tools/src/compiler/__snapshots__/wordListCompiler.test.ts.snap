// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate the wordListCompiler a simple hunspell dictionary depth 0 1`] = `
{
  "error": "",
  "log": "",
}
`;

exports[`Validate the wordListCompiler a simple hunspell dictionary depth 1 1`] = `
{
  "error": "",
  "log": "",
}
`;

exports[`Validate the wordListCompiler compiling to a file without split 1`] = `
{
  "error": "",
  "log": "",
}
`;

exports[`Validate the wordListCompiler reading and normalizing to cities.trie 1`] = `
{
  "error": "",
  "log": "Reading Words into Trie
Reduce duplicate word endings
Writing to file cities.trie
Done writing to file cities.trie",
}
`;

exports[`Validate the wordListCompiler reading and normalizing to cities.trie.gz 1`] = `
{
  "error": "",
  "log": "Reading Words into Trie
Reduce duplicate word endings
Writing to file cities.trie.gz
Done writing to file cities.trie.gz",
}
`;

exports[`Validate the wordListCompiler reading and normalizing to text file: cities.txt 1`] = `
{
  "error": "",
  "log": "",
}
`;

exports[`Validate the wordListCompiler reading and normalizing to text file: cities.txt.gz 1`] = `
{
  "error": "",
  "log": "",
}
`;

exports[`Validate the wordListCompiler tests normalized to a trie 1`] = `
{
  "error": "",
  "log": "",
}
`;
